### PR TITLE
Use specific contract types

### PR DESCRIPTION
**Pull request by NicholasDotSol**
closes: #507 

Rather than storing addresses and then casting to the known contract type, it’s better to use the best type available so the compiler can check for type safety.

### CODE REVIEW

File: solidity/contracts/deposit/Deposit.sol [hunk 1]
```diff
@@ -113,10 +118,10 @@ contract Deposit is DepositFactoryAuthority {
...
         uint256 _lotSizeSatoshis
     ) public onlyFactory payable returns (bool) {
-        self.TBTCSystem = _TBTCSystem;
-        self.TBTCToken = _TBTCToken;
-        self.TBTCDepositToken = _TBTCDepositToken;
-        self.FeeRebateToken = _FeeRebateToken;
+        self.tbtcSystem = ITBTCSystem(_TBTCSystem);
+        self.tbtcToken = TBTCToken(_TBTCToken);
+        self.tbtcDepositToken = IERC721(_TBTCDepositToken);
+        self.feeRebateToken = FeeRebateToken(_FeeRebateToken);
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/Deposit.sol]: Can we declare the actual function parameters to have these types?

File: solidity/contracts/deposit/DepositFunding.sol [hunk 2]
```diff
@@ -55,7 +55,7 @@ library DepositFunding {
         uint256 _n,
         uint256 _lotSizeSatoshis
     ) public returns (bool) {
-        TBTCSystem _system = TBTCSystem(_d.TBTCSystem);
+        TBTCSystem _system = TBTCSystem(address(_d.tbtcSystem));
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/DepositFunding.sol]: Why can't we just reference `_d.tbtcSystem` here? It's already properly typed, no?

File: solidity/contracts/deposit/DepositRedemption.sol [hunk 3]
```diff
@@ -31,13 +31,10 @@ library DepositRedemption {
...
     function distributeSignerFee(DepositUtils.Deposit storage _d) internal {
-        address _tbtcTokenAddress = _d.TBTCToken;
-        TBTCToken _tbtcToken = TBTCToken(_tbtcTokenAddress);
-
         IBondedECDSAKeep _keep = IBondedECDSAKeep(_d.keepAddress);
 
-        _tbtcToken.approve(_d.keepAddress, _d.signerFee());
-        _keep.distributeERC20Reward(_tbtcTokenAddress, _d.signerFee());
+        _d.tbtcToken.approve(_d.keepAddress, _d.signerFee());
+        _keep.distributeERC20Reward(address(_d.tbtcToken), _d.signerFee());
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/DepositRedemption.sol]: This feels like it should propagate to keep, insofar as `distributeERC20Reward` should take an `IERC20`?

File: solidity/contracts/deposit/DepositUtils.sol [hunk 4]
```diff
@@ -354,8 +350,7 @@ library DepositUtils {
     /// @dev            We cast the address to a uint256 to match the 721 standard.
     /// @return         The current deposit beneficiary.
     function depositOwner(Deposit storage _d) public view returns (address payable) {
-        IERC721 _tbtcDepositToken = IERC721(_d.TBTCDepositToken);
-        return address(uint160(_tbtcDepositToken.ownerOf(uint256(address(this)))));
+        return address(uint160(_d.tbtcDepositToken.ownerOf(uint256(address(this)))));
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/DepositUtils.sol]: Is there room here for an overload of `ownerAddressOf` that takes a `Deposit` and returns an `address` and does all of the weird type conversions for us as needed?

File: solidity/contracts/system/TBTCSystem.sol [hunk 5]
```diff
@@ -48,9 +48,9 @@ contract TBTCSystem is Ownable, ITBTCSystem, DepositLog {
     uint256 constant pausedDuration = 10 days;
 
     TBTCDepositToken tbtcDepositToken;
-    address public keepVendor;
-    address public priceFeed;
-    address public relay;
+    IBTCETHPriceFeed priceFeed;
+    IBondedECDSAKeepVendor keepVendor;
+    IRelay relay;
```
Comment (Shadowfiend) [inline solidity/contracts/system/TBTCSystem.sol]: These were declared `public` before, we should probably keep that explicit declaration? Or does leaving off `public` do something else? (Sorry, occasional Solidity gaps exist.)

File: solidity/contracts/deposit/Deposit.sol (hunk 1 above)
Comment (NicholasDotSol) [reply solidity/contracts/deposit/Deposit.sol]: Yes we can, It doesn't really make a difference though

Comment (NicholasDotSol) [reply solidity/contracts/deposit/Deposit.sol]: Will change anyway, might look a little cleaner

File: solidity/contracts/deposit/DepositFunding.sol (hunk 2 above)
Comment (NicholasDotSol) [reply solidity/contracts/deposit/DepositFunding.sol]: _d.tbtcSystem is an `ITBTCSystem` instance. It doesn't have functions like `getAllowNewDeposits()`

Comment (NicholasDotSol) [reply solidity/contracts/deposit/DepositFunding.sol]: It's worth adding necessary functions to the interface actually. In this PR if that's okay with you @Shadowfiend

File: solidity/contracts/deposit/DepositRedemption.sol (hunk 3 above)
Comment (NicholasDotSol) [reply solidity/contracts/deposit/DepositRedemption.sol]: Don't think I follow :s

File: solidity/contracts/deposit/DepositUtils.sol (hunk 4 above)
Comment (NicholasDotSol) [reply solidity/contracts/deposit/DepositUtils.sol]: Yeah that might be a good idea.

File: solidity/contracts/system/TBTCSystem.sol (hunk 5 above)
Comment (NicholasDotSol) [reply solidity/contracts/system/TBTCSystem.sol]: > Or does leaving off public do something else

Nope, it's just public by default. I'll add the visibility back for consistency.

File: solidity/contracts/deposit/DepositFunding.sol (hunk 2 above)
Comment (Shadowfiend) [reply solidity/contracts/deposit/DepositFunding.sol]: Yes please, awesome plan!

File: solidity/contracts/deposit/DepositRedemption.sol (hunk 3 above)
Comment (Shadowfiend) [reply solidity/contracts/deposit/DepositRedemption.sol]: Sorry, I mean `distributeERC20Reward` should declare its first parameter as an `IERC20` rather than an `address`, and then we wouldn't need this cast, right? Obviously we'd have to make that change in keep-ecdsa as well.

File: solidity/contracts/deposit/DepositUtils.sol (hunk 4 above)
Comment (Shadowfiend) [reply solidity/contracts/deposit/DepositUtils.sol]: Probably not worth cycles unless you're motivated, but yeah, seems cleaner.

File: solidity/contracts/proxy/DepositFactory.sol [hunk 6]
```diff
@@ -51,9 +53,9 @@ contract DepositFactory is CloneFactory, TBTCSystemAuthority{
     ) public onlyTbtcSystem {
         masterDepositAddress = _masterDepositAddress;
         tbtcDepositToken = TBTCDepositToken(_tbtcDepositToken);
-        tbtcSystem = _tbtcSystem;
-        tbtcToken = _tbtcToken;
-        feeRebateToken = _feeRebateToken;
+        tbtcSystem = TBTCSystem(_tbtcSystem);
```
Comment (Shadowfiend) [inline solidity/contracts/proxy/DepositFactory.sol]: Sort of in the same vein, can we bubble this typing up to the parameters for `setExternalDependencies` as well?

File: solidity/contracts/proxy/DepositFactory.sol [hunk 7]
```diff
@@ -19,9 +21,9 @@ contract DepositFactory is CloneFactory, TBTCSystemAuthority{
     // which will be used as a master contract for cloning.
     address payable public masterDepositAddress;
     TBTCDepositToken tbtcDepositToken;
-    address public tbtcSystem;
-    address public tbtcToken;
-    address public feeRebateToken;
+    TBTCSystem public tbtcSystem;
+    TBTCToken public tbtcToken;
+    FeeRebateToken public feeRebateToken;
     address public vendingMachine;
```
Comment (Shadowfiend) [inline solidity/contracts/proxy/DepositFactory.sol]: Should we hit this one?

File: solidity/contracts/interfaces/ITBTCSystem.sol [hunk 8]
```diff
@@ -16,4 +16,10 @@ interface ITBTCSystem {
     function fetchRelayPreviousDifficulty() external view returns (uint256);
 
     function getInitialCollateralizedPercent() external view returns (uint128);
+    function getAllowNewDeposits() external view returns (bool);
+    function isAllowedLotSize(uint256 _lotSizeSatoshis) external view returns (bool);
+    function requestNewKeep(uint256 _m, uint256 _n, uint256 _bond) external payable returns (address);
+    function getSignerFeeDivisor() external view returns (uint256);
+    function getUndercollateralizedThresholdPercent() external view returns (uint128);
+    function getSeverelyUndercollateralizedThresholdPercent() external view returns (uint128);
```
Comment (Shadowfiend) [inline solidity/contracts/interfaces/ITBTCSystem.sol]: Arguably our docs should be here rather than on `TBTCSystem`, but I need to see how our documentation generator is operating.

File: solidity/contracts/proxy/DepositFactory.sol (hunk 7 above)
Comment (NicholasDotSol) [reply solidity/contracts/proxy/DepositFactory.sol]: We only use the address and never call any `VendingMachine` contract functions.
Hitting this would be counterproductive since we would just have to cast it back to use it.

Comment (Shadowfiend) [reply solidity/contracts/proxy/DepositFactory.sol]: Why do we use the address? `createNewDeposit` should likewise take a `VendingMachine`, right? Generally eschewing the type check is a bad idea, and if we have to cast to a looser type at the use point, that's fine---otherwise you're leaving on the table the capacity of the compiler to bonk you when you mishandle some point in the initialization chain.

Comment (NicholasDotSol) [reply solidity/contracts/proxy/DepositFactory.sol]: > Why do we use the address? 

It's just used for redemption logic:
` if(tdtHolder == vendingMachine){...}`

Worth noting, It's introduced into the system as a contract type via tbtcSystem.initialize().
Then it's passed along and stored as an address. Might be worth renaming it to `vendingMachineAddress` to avoid confusion?

File: solidity/contracts/deposit/DepositRedemption.sol (hunk 3 above)
Comment (NicholasDotSol) [reply solidity/contracts/deposit/DepositRedemption.sol]: I'm not sure about this. At this point we're losing any compiler advantages anyway, no?

File: solidity/contracts/proxy/DepositFactory.sol (hunk 7 above)
Comment (Shadowfiend) [reply solidity/contracts/proxy/DepositFactory.sol]: I like this. Makes it clear that we're really only interested in the address.

Also makes it clear this is a dirty dirty hack and in another life where we had more time we'd try to get around it haha.

File: solidity/contracts/deposit/DepositRedemption.sol (hunk 3 above)
Comment (Shadowfiend) [reply solidity/contracts/deposit/DepositRedemption.sol]: Hmmm… How so? The interface would require `IERC20`, and the Keep side also implements that interface. Even though there isn't a build of the whole system where the compiler will catch stuff, the interfaces allow us to catch those same issues at compile time.

File: solidity/contracts/system/TBTCSystem.sol [hunk 9]
```diff
@@ -79,41 +81,41 @@ contract TBTCSystem is Ownable, ITBTCSystem, DepositLog {
...
+        _vendingMachine.setExternalAddresses(
             _tbtcToken,
             _tbtcDepositToken,
             _feeRebateToken
         );
-        DepositFactory(_depositFactory).setExternalDependencies(
+        _depositFactory.setExternalDependencies(
             _masterDepositAddress,
-            address(this),
+            TBTCSystem(address(this)),
```
Comment (Shadowfiend) [inline solidity/contracts/system/TBTCSystem.sol]: Can this not just be `this`?

File: solidity/contracts/proxy/DepositFactory.sol (hunk 7 above)
Comment (NicholasDotSol) [reply solidity/contracts/proxy/DepositFactory.sol]: ok awesome

File: solidity/contracts/deposit/DepositRedemption.sol (hunk 3 above)
Comment (NicholasDotSol) [reply solidity/contracts/deposit/DepositRedemption.sol]: Alright, let's change it. Needs `keep-ecdsa` PR.

File: solidity/contracts/system/TBTCSystem.sol (hunk 9 above)
Comment (NicholasDotSol) [reply solidity/contracts/system/TBTCSystem.sol]: it can, and it should

File: solidity/contracts/deposit/Deposit.sol [hunk 10]
```diff
@@ -122,7 +122,7 @@ contract Deposit is DepositFactoryAuthority {
         self.tbtcToken = _TBTCToken;
         self.tbtcDepositToken = _TBTCDepositToken;
         self.feeRebateToken = _FeeRebateToken;
-        self.VendingMachine = _VendingMachine;
+        self.VendingMachineAddress = _VendingMachineAddress;
```
Comment (Shadowfiend) [inline solidity/contracts/deposit/Deposit.sol]: … Feels like all of these should start with a lowercase? And does that maybe also include all the `_feeRebateToken`, `_tbtcDepositToken`, etc?

File: solidity/contracts/test/deposit/TestDeposit.sol [hunk 11]
```diff
@@ -36,13 +36,13 @@ contract TestDeposit is Deposit {
...
         address _feeRebateToken,
-        address _vendingMachine
+        address _vendingMachineAddress
     ) public {
         self.tbtcSystem = ITBTCSystem(_sys);
         self.tbtcToken = TBTCToken(_token);
         self.tbtcDepositToken = IERC721(_tbtcDepositToken);
         self.feeRebateToken = FeeRebateToken(_feeRebateToken);
-        self.VendingMachine = _vendingMachine;
+        self.VendingMachineAddress = _vendingMachineAddress;
```
Comment (Shadowfiend) [inline solidity/contracts/test/deposit/TestDeposit.sol]: Ditto on the caps.

File: solidity/contracts/test/keep/ECDSAKeepStub.sol [hunk 12]
```diff
@@ -35,7 +36,7 @@ contract ECDSAKeepStub is IBondedECDSAKeep {
         // solium-disable-previous-line no-empty-blocks
     }
 
-    function distributeERC20Reward(address _asset, uint256 _value) external {
+    function distributeERC20Reward(IERC20 _asset, uint256 _value) external {
```
Comment (Shadowfiend) [inline solidity/contracts/test/keep/ECDSAKeepStub.sol]: This won't be enough, we have to play this out over both repos. Let's do that in a separate PR.

Comment (NicholasDotSol) [reply solidity/contracts/test/keep/ECDSAKeepStub.sol]: > This won't be enough

We just need an interface change from `keep-ecdsa`.  This should be enough for TBTC

File: solidity/contracts/deposit/Deposit.sol (hunk 10 above)
Comment (NicholasDotSol) [reply solidity/contracts/deposit/Deposit.sol]: I can go either way, as long as we're consistent. 
Changing to lowercase for consistency

File: solidity/contracts/test/keep/ECDSAKeepStub.sol (hunk 12 above)
Comment (NicholasDotSol) [reply solidity/contracts/test/keep/ECDSAKeepStub.sol]: > we have to play this out over both repos

Okay with moving to separate PR to unblock

Comment (Shadowfiend) [reply solidity/contracts/test/keep/ECDSAKeepStub.sol]: Well, the interface change will also require changing the implementors of the interface, etc. It's not going to be huge, just needs a little more finesse :)

File: solidity/contracts/deposit/DepositRedemption.sol (hunk 3 above)
Comment (NicholasDotSol) [reply solidity/contracts/deposit/DepositRedemption.sol]: moving to separate PR
